artifact.go

```diff
@@ -163,50 +151,311 @@ func (m ArtifactModule) getBuildTestsResultsHandler(c *gin.Context) {
+// @router /build/{buildid}/test-result-details [get]
+func (m ArtifactModule) getBuildAllTestResultDetailsHandler(c *gin.Context) {
+	buildID, ok := parseBuildID(c)
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: uint(buildID)}).
+		Find(&details).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result details for build with ID %d from database.",
+			buildID))
+		return
+	}
+
+	c.JSON(http.StatusOK, details)
+}
+
+// getBuildTestResultDetailsHandler godoc
+// @summary Get test result details for specified test
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @param artifactId path int true "Artifact ID"
+// @success 200 {object} []TestResultDetail
+// @router /build/{buildid}/test-result-details/{artifactId} [get]
+func (m ArtifactModule) getBuildTestResultDetailsHandler(c *gin.Context) {
+	buildID, artifactID, ok := parseBuildAndArtifactID(c)
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: uint(buildID), ArtifactID: artifactID}).
+		Find(&details).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result details from test with ID %d for build with ID %d from database.",
+			artifactID, buildID))
+		return
+	}
+
+	c.JSON(http.StatusOK, details)
+}
+
+// getBuildTestResultsSummaryHandler godoc
+// @summary Get test result summary of all tests for specified build
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @success 200 {object} SummaryOfTestResultSummaries
+// @failure 400 {object} problem.Response "Bad Request"
+// @failure 502 {object} problem.Response "Bad Gateway"
+// @router /build/{buildid}/test-results-summary [get]
+func (m ArtifactModule) getBuildTestResultsSummaryHandler(c *gin.Context) {
+	buildID, ok := parseBuildID(c)
 	if !ok {
 		return
 	}
 
+	summaries := []TestResultSummary{}
+	err := m.Database.
+		Preload("Artifact").
+		Where(&TestResultSummary{BuildID: uint(buildID)}).
+		Find(&summaries).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result summaries for build with ID %d from database.",
+			buildID))
+		return
+	}
+
+	summary := SummaryOfTestResultSummaries{
+		BuildID:   buildID,
+		Summaries: summaries}
+
+	for _, v := range summaries {
+		summary.Failed += v.Failed
+		summary.Passed += v.Passed
+		summary.Skipped += v.Skipped
+	}
+
+	summary.Total = summary.Failed + summary.Passed + summary.Skipped
+
+	c.JSON(http.StatusOK, summary)
+}
+
+// parseMultipartFormData writes 400 "Bad request" problem.Response on failure.
+// Returns a slice of file pointers on success, or an empty slice if there were
+// none but parsing was successful.
+func parseMultipartFormData(c *gin.Context) ([]*file, bool) {
+	buildID, ok := parseBuildID(c)
+	if !ok {
+		return nil, false
+	}
```

Comment:
Dependency on the `{buildid}` parameter is quite hidden in this function. I suggest to accept it through a parameter or remove it from the problem detail strings